pubsub/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/PubSubSink.scala

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2013-2023 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.collectors.scalastream.sinks
+
+import cats.Parallel
+import cats.effect.implicits.genSpawnOps
+import cats.effect.{Async, Ref, Resource, Sync}
+import cats.implicits._
+import com.google.api.gax.retrying.RetrySettings
+import com.google.api.gax.rpc.{ApiException, FixedHeaderProvider}
+import com.permutive.pubsub.producer.Model.{ProjectId, Topic}
+import com.permutive.pubsub.producer.encoder.MessageEncoder
+import com.permutive.pubsub.producer.grpc.{GooglePubsubProducer, PubsubProducerConfig}
+import com.permutive.pubsub.producer.{Model, PubsubProducer}
+import com.snowplowanalytics.snowplow.collector.core.{Config, Sink}
+import com.snowplowanalytics.snowplow.collectors.scalastream.BuildInfo
+import com.snowplowanalytics.snowplow.collectors.scalastream.sinks.BuilderOps._
+import org.threeten.bp.Duration
+import org.typelevel.log4cats.Logger
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+import retry.RetryPolicies
+import retry.syntax.all._
+
+import scala.concurrent.duration.{DurationLong, FiniteDuration}
+import scala.util._
+
+class PubSubSink[F[_]: Async: Parallel: Logger] private (
+  override val maxBytes: Int,
+  isHealthyState: Ref[F, Boolean],
+  producer: PubsubProducer[F, Array[Byte]],
+  retryInterval: FiniteDuration,
+  topicName: String
+) extends Sink[F] {
+
+  override def storeRawEvents(events: List[Array[Byte]], key: String): F[Unit] =
+    produceBatch(events).start.void
+
+  override def isHealthy: F[Boolean] = isHealthyState.get
+
+  private def produceBatch(events: List[Array[Byte]]): F[Unit] =
+    events.parTraverse_ { event =>
+      produceSingleEvent(event) *> isHealthyState.set(true)
+    }
+
+  private def produceSingleEvent(event: Array[Byte]): F[Model.MessageId] =
+    producer
+      .produce(event)
+      .retryingOnAllErrors(
+        policy  = RetryPolicies.constantDelay(retryInterval),
+        onError = (error, _) => handlePublishError(error)
+      )
+
+  private def handlePublishError(error: Throwable): F[Unit] =
+    isHealthyState.set(false) *> Logger[F].error(createErrorMessage(error))
```

Comment:
Related to the discussion about logging policy, we'll need to decide if we should be logging errors for individual events or if we need some batching.

Reply:
Good point.  This only becomes an issue if there is an outage and if it lasts for a long time.  Ideally we should find a nice pattern in the code for de-duping log messages upon errors.  And we should start using it everywhere when we've decided on that pattern.

No need to hold up this current PR though.